cpp/src/quantiles/quantiles.cu

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <thrust/functional.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <cudf/copying.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <quantiles/quantiles_util.hpp>
+
+using ScalarResult = double;
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+namespace {
+
+struct quantile_functor
+{
+    template<typename T>
+    std::enable_if_t<not std::is_arithmetic<T>::value, std::unique_ptr<scalar>>
+    operator()(column_view const& input,
+               double q,
+               interpolation interp,
+               bool is_sorted,
+               cudf::order col_order,
+               cudf::null_order col_null_order,
+               rmm::mr::device_memory_resource *mr =
+               rmm::mr::get_default_resource(),
+               cudaStream_t stream = 0)
```

Comment:
Wanna see a trick I use to clean up the case when you just have to throw an exception? see : https://github.com/rapidsai/cudf/blob/995c5100423f2d3e2c5b61114800c93ea3cd4069/cpp/src/reductions/legacy/group_std.cu#L113-L117